### PR TITLE
Implement motion vectors and TAA for skinned meshes and meshes with morph targets.

### PR DESCRIPTION
**Pull request by pcwalton**
This is a revamped equivalent to #9902, though it shares none of the code. It handles all special cases that I've tested correctly.

The overall technique consists of double-buffering the joint matrix and morph weights buffers, as most of the previous attempts to solve this problem did. The process is generally straightforward. Note that, to avoid regressing the ability of mesh extraction, skin extraction, and morph target extraction to run in parallel, I had to add a new system to rendering, `set_mesh_motion_vector_flags`. The comment there explains the details; it generally runs very quickly.

I've tested this with modified versions of the `animated_fox`, `morph_targets`, and `many_foxes` examples that add TAA, and the patch works. To avoid bloating those examples, I didn't add switches for TAA to them.

Addresses points (1) and (2) of #8423.

### Fixed

* Motion vectors, and therefore TAA, are now supported for meshes with skins and/or morph targets.

### CODE REVIEW

**Review by aevyrie — APPROVED** [review]
Closes an important gap in bevy, very little added code, this is largely documentation - the code quality here is great. The simplifying assumptions here w.r.t. the buffers seems well justified. All-in-all this should be pretty uncontroversial.

File: crates/bevy_pbr/src/render/morph.rs [hunk 1]
```diff
@@ -80,25 +107,32 @@ fn add_to_alignment<T: NoUninit + Default>(buffer: &mut RawBufferVec<T>) {
...
+    morph_indices: ResMut<MorphIndices>,
+    uniform: ResMut<MorphUniforms>,
     query: Extract<Query<(Entity, &ViewVisibility, &MeshMorphWeights)>>,
 ) {
-    morph_indices.clear();
-    uniform.buffer.clear();
+    // Borrow check workaround.
+    let (morph_indices, uniform) = (morph_indices.into_inner(), uniform.into_inner());
+
+    // Swap buffers. We need to keep the previous frame's buffer around for the
```
Comment (JMS55) [inline crates/bevy_pbr/src/render/morph.rs:117]: I think your way is a bit cleaner, but this is how I did it in my PR https://github.com/JMS55/bevy/blob/3334d0bd5ff643f54a605672a70d05b6b271b455/crates/bevy_pbr/src/render/morph.rs#L93-L110. Can't remember why they look different off the top of my head.

**Review by JMS55 — CHANGES_REQUESTED** [review]
LGTM minus some general nits, and pending tracy comparisons on a large many_foxes run.

EDIT: Please also update the doc comments in taa/mod.rs talking about skinned/morph support.

**Comment by pcwalton** [comment]
This is approximately a 3%-4% regression in CPU time of `main_opaque_pass_3d` of `many_foxes` with 10,000 foxes.

CPU 13th Gen Intel(R) Core(TM) i9-13900H
GPU NVIDIA GeForce RTX 4070 Laptop GPU
Screenshot 2024-05-29 225937

Mac mini, M2
Screenshot 2024-05-29 231740

I'd argue that this small regression is worth it.

File: crates/bevy_pbr/src/render/mesh.rs [hunk 2]
```diff
@@ -1896,18 +1969,34 @@ pub fn prepare_mesh_bind_group(
...
     groups.model_only = Some(layouts.model_only(&render_device, &model));
 
-    let skin = skins_uniform.buffer.buffer();
+    // Create the skinned mesh bind group with the current and previous buffers
+    // (the latter being for motion vector computation). If there's no previous
+    // buffer, just use the current one as the shader will ignore it.
+    let skin = skins_uniform.current_buffer.buffer();
     if let Some(skin) = skin {
-        groups.skinned = Some(layouts.skinned(&render_device, &model, skin));
+        let prev_skin = skins_uniform.prev_buffer.buffer().unwrap_or(skin);
```
[the remark this answers is not in the document]
Comment (pcwalton) [reply crates/bevy_pbr/src/render/mesh.rs:1977]: I didn't think there was any kind of rule against binding a read-only buffer twice. I can see problems with read-write buffers, but read-only seems always fine. It's extremely common for textures, so I don't see why it wouldn't be for buffers...

File: crates/bevy_pbr/src/prepass/prepass.wgsl [hunk 3]
```diff
@@ -93,12 +112,42 @@ fn vertex(vertex_no_morph: Vertex) -> VertexOutput {
     out.color = vertex.color;
 #endif
 
+    // Compute the motion vector, for TAA. For this we need to know where the
+    // vertex was last frame.
 #ifdef MOTION_VECTOR_PREPASS
-    // Use vertex_no_morph.instance_index instead of vertex.instance_index to work around a wgpu dx12 bug.
```
[the remark this answers is not in the document]
Comment (pcwalton) [reply crates/bevy_pbr/src/prepass/prepass.wgsl:97]: I just tried with the Direct3D 12 backend and it works fine. I guess whatever the issue was is now fixed.

File: crates/bevy_pbr/src/prepass/prepass.wgsl [hunk 4]
```diff
@@ -93,12 +112,42 @@ fn vertex(vertex_no_morph: Vertex) -> VertexOutput {
     out.color = vertex.color;
 #endif
 
+    // Compute the motion vector, for TAA. For this we need to know where the
```
[the remark this answers is not in the document]
Comment (alice-i-cecile) [reply crates/bevy_pbr/src/prepass/prepass.wgsl:115]:
```suggestion
    // Compute the motion vector, e.g. for TAA. For this we need to know where the
```

File: crates/bevy_pbr/src/render/mesh.rs [hunk 5]
```diff
@@ -412,12 +411,18 @@ bitflags::bitflags! {
         const AUTOMATIC_BATCHING      = 1 << 1;
         /// The mesh had a transform last frame and so is eligible for TAA.
```
[the remark this answers is not in the document]
Comment (pcwalton) [reply crates/bevy_pbr/src/render/mesh.rs]: Done

File: crates/bevy_pbr/src/render/mesh.rs [hunk 6]
```diff
@@ -412,12 +411,18 @@ bitflags::bitflags! {
...
         const HAVE_PREVIOUS_TRANSFORM = 1 << 2;
```
[the remark this answers is not in the document]
Comment (pcwalton) [reply crates/bevy_pbr/src/render/mesh.rs]: Renamed

File: crates/bevy_pbr/src/prepass/prepass.wgsl (hunk 4 above)
[the remark this answers is not in the document]
Comment (pcwalton) [reply crates/bevy_pbr/src/prepass/prepass.wgsl]: I said "for TAA among other purposes"

**Comment by pcwalton** [comment]
The regression is somewhat worse--13% or so for `main_opaque_pass_3d`--on my slower box:
CPU: 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz
GPU: NVIDIA GeForce RTX 2060
Screenshot 2024-05-30 214741

Frame time is a 5.73% regression: 
Screenshot 2024-05-30 215220

I still think we should take this regression as this is still not very large and this is a stress test. Most games won't see this kind of regression.

**Comment by pcwalton** [comment]
This version of the patch should eliminate the regression by only including the bindings for views with motion vectors.

**Review by JMS55 — APPROVED** [review]
I think we're technically doing a double pipeline compile here, one the first frame you add a skinned entity when it doesn't yet have a previous skin buffer entry, and then a second pipeline the next frame when it does.

I'm ok with that though, we can always improve it later.

**Comment by alice-i-cecile** [comment]
Thank you to everyone involved with the authoring or reviewing of this PR! This work is relatively important and needs release notes! Head over to https://github.com/bevyengine/bevy-website/issues/1326 if you'd like to help out.

**Pull request merged**